src/backend/distributed/utils/metadata_cache.c

```diff
@@ -1483,13 +1698,43 @@ ResetDistTableCacheEntry(DistTableCacheEntry *cacheEntry)
 
 	if (cacheEntry->shardIntervalArrayLength > 0)
 	{
-		int i = 0;
+		int shardIndex = 0;
 
-		for (i = 0; i < cacheEntry->shardIntervalArrayLength; i++)
+		for (shardIndex = 0; shardIndex < cacheEntry->shardIntervalArrayLength;
+			 shardIndex++)
 		{
-			ShardInterval *shardInterval = cacheEntry->sortedShardIntervalArray[i];
+			ShardInterval *shardInterval =
+				cacheEntry->sortedShardIntervalArray[shardIndex];
 			bool valueByVal = shardInterval->valueByVal;
+			bool foundInCache = false;
+			int placementIndex = 0;
+			ShardPlacement *placementArray =
+				cacheEntry->arrayOfPlacementArrays[shardIndex];
+			int numberOfPlacements =
+				cacheEntry->arrayOfPlacementArrayLengths[shardIndex];
+
+			/* delete the shard's placements */
+			for (placementIndex = 0;
+				 placementIndex < numberOfPlacements;
+				 placementIndex++)
+			{
+				ShardPlacement *placement = &placementArray[placementIndex];
+
+				if (placement->nodeName)
+				{
+					pfree(placement->nodeName);
+				}
 
+				/* placement itself is deleted as part of the array */
+			}
+			pfree(placementArray);
```

Comment:
This function is getting huge. Is this something we can do in a MemoryContext? (might be more effort than it's worth)